daemon/hooks/hooks.go

```diff
@@ -13,25 +13,55 @@ type Hook interface {
 	Filesystems() Filter
 	ErrIsFatal() bool
 	AlwaysRun() bool
-	Run(ctx context.Context, edge HookEdge, phase string, dryRun bool, extra HookEnv) RunError
+	Run(ctx context.Context, edge HookEdge, phase HookPhase, dryRun bool, extra HookEnv) HookReport
 	Disable()
+	String() string
+}
+
+type HookPhase string
+
+const (
+	PhaseSnapshot = HookPhase("snapshot")
+	PhaseTesting  = HookPhase("testing")
+)
+
+func (p HookPhase) String() string {
+	return string(p)
 }
 
 //go:generate stringer -type=HookEdge
 type HookEdge uint
 
+// Following order MUST be correct run
+// order, as is used by Hook.Before() method
+// to sort hooks
 const (
-	Disabled = HookEdge(0)
-	All      = HookEdge(^uint(0)) // max value, all bits == 1
-	Pre      = HookEdge(1 << iota)
+	Pre = HookEdge(1 << iota)
+	Callback
 	Post
 	Cleanup
+	// "Special" values must be last, or
+	// iota gets out of sequence
+	Disabled = HookEdge(0)
+	All      = HookEdge(^uint(0)) // max value, all bits == 1
 )
 
 func (e HookEdge) RunsDuring(runningEdge HookEdge) bool {
 	return e&runningEdge != 0
 }
 
+func (e HookEdge) StringForPhase(phase HookPhase) string {
+	// Cleanup is a special type of Post-hook,
```

Comment:
`special`: someone is going to scratch their head about this in two years. Refer to the place where you explain how config semantics are translated to the edge semantics. If such a section doesn't exist, introduce one in a `hook_docs.go` package-level comment or where the Edge constants are defined.

Reply:
With the dependency planner rewrite, `Cleanup` is now gone.
### Before
* `Hook.Edge() & Post > 0 && Hook.AlwaysRun() == true`
  ≡
  `hookPlanEntry.Edge == Cleanup`
* Sort by
  1. `hookPlanEntry.Edge`
  2. `HookJobPlan.steps` slice order
### After
* No `Cleanup` edge
* Sort by
  1. `hookPlanEntry.Edge`
  2. `Hook.AlwaysRun() == true` > `Hook.AlwaysRun() == false`
  3. `hookPlanEntry.configOrder`